File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        Stack<Integer> st = new Stack<>();
        int trapWater = 0, n = height.length;
        for(int i = 0; i < n; i++) {
            int prevH = 0;
            int curr = i;
            while(!st.isEmpty() && height[st.peek()] <= height[i]) {
                curr = st.pop();
                trapWater += (i - curr - 1) * (height[curr] - prevH);
                prevH = height[curr];
            }
            if(!st.isEmpty()) {
                trapWater += (i - st.peek() - 1) * (height[i] - prevH);
            }
            st.push(i);
        }
        return trapWater;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> mono_stack;
        int count = 0, idx = 0, i, prev = 0;
        for(auto h:height){
            while(!mono_stack.empty() && height[mono_stack.top()] <= h){
                i = mono_stack.top();
                count += (height[i] - prev)* (idx - i - 1);
                prev = height[i];
                mono_stack.pop();
            }
            
            if(!mono_stack.empty()){
                count += (h - prev) * (idx - mono_stack.top() - 1);
            }
            
            mono_stack.push(idx);
            prev = h;
            idx++;
        }
        return count;
    }
};